src/main/scala/Day07Copy.scala

```diff
@@ -0,0 +1,74 @@
+import scala.io.*
+
+object Day07Copy extends App:
+
+  private val day: String =
+    this.getClass.getName.drop(3).init
+
+  val start1: Long = System.currentTimeMillis
+  val start2: Long = System.currentTimeMillis
+
+  private val rules: List[Rule] =
+    Source
+      .fromResource(s"input$day.txt")
+      .getLines
+      .toList
+      .map {
+        case s"Step ${before} must be finished before step ${after} can begin." => Rule(before.head, after.head)
+      }
+
+  case class Rule(before: Char, after: Char)
+
+  def createLetterSet(r: List[Rule]): Set[Char] = {
+    def go(r: List[Rule], s: Set[Char])(f: Rule => Char) : Set[Char] = {
+      r match
+        case Nil => s
+        case h :: t => go(t, s + f(h))(f)
+    }
+    go(r, go(r, Set.empty)(e => e.before))(e => e.after)
+  }
+
+  val letterSet = createLetterSet(rules)
+
+  def setOrder(r: List[Rule], s: Set[Char]): List[Char] = {
+    def go(r: List[Rule], s: Set[Char], order: List[Char]): List[Char] = {
+      if(s.isEmpty) order
+      else {
+        val available = s.filterNot(ss => r.map(_.after).contains(ss)).min
+        go(r.filterNot(e => e.before == available), s - available, available :: order)
```

Comment:
could use `vector :+ element` to append instead of list

( works also on list - but slow )